### PR TITLE
fix score indices of fzf-native when using multi bytes

### LINKED ISSUE #14: fussy-socre causes an args-out-of-range error when using fussy-fzf-native-score and passing non-ascii characters
The repoducable code is here:

```lisp
(let ((fussy-filter-fn #'fussy-filter-flex)
      (fussy-score-fn #'fussy-fzf-native-score))
  (fussy-score '("ポケモン.txt") "txt"))
;; => error: (args-out-of-range 13 16)
```

It seems that the position returned by `fussy-fzf-native` is calculated as a byte string instead of a string, so `fussy-propertize-common-part` results in `args-out-of-range`.

```
(fussy-fzf-native-score "ポケモン.txt" "txt")
;; => (80 13 14 15)
(fussy-fzf-native-score "pckemon.txt" "txt")
;; => (80 8 9 10)
```

I think this is a fzf-native issue, but I would be happy to have a workaround because fzf-native does not seem to take care of non-ascii (e.g. [here](https://github.com/dangduc/fzf-native/blob/506f3bd20fbf9860dd913a09b53661a27e4edc42/fzf.c#L369-L377)).

## Backtrace
<details>

```
Debugger entered--Lisp error: (args-out-of-range 13 16)
  add-face-text-property(13 16 completions-common-part nil #("ポケモン.txt" 0 1 (completion-score 80)))
  (let ((block-started (car (cdr score))) (last-char nil) (str (if (consp obj) (car obj) obj))) (let ((--dolist-tail-- (cdr score))) (while --dolist-tail-- (let ((char (car --dolist-tail--))) (if (and last-char (not (= ... char))) (progn (add-face-text-property block-started (1+ last-char) 'completions-common-part nil str) (setq block-started char))) (setq last-char char) (setq --dolist-tail-- (cdr --dolist-tail--))))) (add-face-text-property block-started (1+ last-char) 'completions-common-part nil str) (if (and last-char (> (length str) (+ 2 last-char))) (progn (add-face-text-property (1+ last-char) (+ 2 last-char) 'completions-first-difference nil str))) (if (consp obj) (cons str (cdr obj)) str))
  (if (<= (length score) 1) obj (let ((block-started (car (cdr score))) (last-char nil) (str (if (consp obj) (car obj) obj))) (let ((--dolist-tail-- (cdr score))) (while --dolist-tail-- (let ((char (car --dolist-tail--))) (if (and last-char (not ...)) (progn (add-face-text-property block-started ... ... nil str) (setq block-started char))) (setq last-char char) (setq --dolist-tail-- (cdr --dolist-tail--))))) (add-face-text-property block-started (1+ last-char) 'completions-common-part nil str) (if (and last-char (> (length str) (+ 2 last-char))) (progn (add-face-text-property (1+ last-char) (+ 2 last-char) 'completions-first-difference nil str))) (if (consp obj) (cons str (cdr obj)) str)))
  fussy-propertize-common-part(#("ポケモン.txt" 0 1 (completion-score 80)) (80 13 14 15))
  funcall(fussy-propertize-common-part #("ポケモン.txt" 0 1 (completion-score 80)) (80 13 14 15))
  (setq x (funcall fussy-propertize-fn x score))
  (progn (setq x (funcall fussy-propertize-fn x score)))
  (if (fussy--should-propertize-p) (progn (setq x (funcall fussy-propertize-fn x score))))
  (progn (put-text-property 0 1 'completion-score (car score) x) (if (fussy--should-propertize-p) (progn (setq x (funcall fussy-propertize-fn x score)))) (setq result (cons x result)))
  (if (and score (> (car score) 0)) (progn (put-text-property 0 1 'completion-score (car score) x) (if (fussy--should-propertize-p) (progn (setq x (funcall fussy-propertize-fn x score)))) (setq result (cons x result))))
  (let ((score (funcall fussy-score-fn x string cache))) (if (and score (> (car score) 0)) (progn (put-text-property 0 1 'completion-score (car score) x) (if (fussy--should-propertize-p) (progn (setq x (funcall fussy-propertize-fn x score)))) (setq result (cons x result)))))
  (if (> (length x) fussy-max-word-length-to-score) (setq result (cons x result)) (let ((score (funcall fussy-score-fn x string cache))) (if (and score (> (car score) 0)) (progn (put-text-property 0 1 'completion-score (car score) x) (if (fussy--should-propertize-p) (progn (setq x (funcall fussy-propertize-fn x score)))) (setq result (cons x result))))))
  (let ((x (car --dolist-tail--))) (setq x (copy-sequence x)) (if (> (length x) fussy-max-word-length-to-score) (setq result (cons x result)) (let ((score (funcall fussy-score-fn x string cache))) (if (and score (> (car score) 0)) (progn (put-text-property 0 1 'completion-score (car score) x) (if (fussy--should-propertize-p) (progn (setq x ...))) (setq result (cons x result)))))) (setq --dolist-tail-- (cdr --dolist-tail--)))
  (while --dolist-tail-- (let ((x (car --dolist-tail--))) (setq x (copy-sequence x)) (if (> (length x) fussy-max-word-length-to-score) (setq result (cons x result)) (let ((score (funcall fussy-score-fn x string cache))) (if (and score (> (car score) 0)) (progn (put-text-property 0 1 'completion-score (car score) x) (if (fussy--should-propertize-p) (progn ...)) (setq result (cons x result)))))) (setq --dolist-tail-- (cdr --dolist-tail--))))
  (let ((--dolist-tail-- candidates)) (while --dolist-tail-- (let ((x (car --dolist-tail--))) (setq x (copy-sequence x)) (if (> (length x) fussy-max-word-length-to-score) (setq result (cons x result)) (let ((score (funcall fussy-score-fn x string cache))) (if (and score (> ... 0)) (progn (put-text-property 0 1 ... ... x) (if ... ...) (setq result ...))))) (setq --dolist-tail-- (cdr --dolist-tail--)))))
  (let ((result 'nil)) (let ((--dolist-tail-- candidates)) (while --dolist-tail-- (let ((x (car --dolist-tail--))) (setq x (copy-sequence x)) (if (> (length x) fussy-max-word-length-to-score) (setq result (cons x result)) (let ((score ...)) (if (and score ...) (progn ... ... ...)))) (setq --dolist-tail-- (cdr --dolist-tail--))))) (reverse result))
  fussy-score(("ポケモン.txt") "txt")
  (let ((fussy-filter-fn #'fussy-filter-flex) (fussy-score-fn #'fussy-fzf-native-score)) (fussy-score '("ポケモン.txt") "txt"))
  (progn (let ((fussy-filter-fn #'fussy-filter-flex) (fussy-score-fn #'fussy-fzf-native-score)) (fussy-score '("ポケモン.txt") "txt")))
  elisp--eval-last-sexp(t)
  eval-last-sexp(t)
  eval-print-last-sexp(nil)
  funcall-interactively(eval-print-last-sexp nil)
  command-execute(eval-print-last-sexp)
```
</details>

### PR DESCRIPTION
**Pull request by buzztaiki**
fixes #14

### CODE REVIEW

File: fussy-test.el [hunk 1]
```diff
@@ -353,8 +353,8 @@ This test asserts `fussy-encode-coding-string' is much much faster than
 `fussy-without-unencodeable-chars'."
   (let* ((tofu (char-to-string fussy--consult--tofu-char))
          (string-1 (concat "jjbb" tofu))
-         (string-2 (string-as-multibyte ";; Copyright 2022 Jo Be�"))
-         (string-3 (string-as-multibyte ";; This buffer is for text that is not saved, and for Lisp evaluation.�"))
+         (string-2 (string-as-multibyte ";; Copyright 2022 Jo Be\370\210\200\201\222"))
```
Comment (jojojames) [inline fussy-test.el:356]: Maybe leave these alone for the test.

File: fussy.el [hunk 2]
```diff
@@ -1082,13 +1082,27 @@ highlighting."
       (when (fboundp 'fzf-native-make-default-slab)
         (setf fussy--fzf-native-slab (fzf-native-make-default-slab)))))
 
+(defun fussy--fzf-native-fix-score-indices (str score)
+  "Fix SCORE indices of `fzf-native' if STR is multibyte."
+  (if (or (not (multibyte-string-p str)) (null score))
```
Comment (jojojames) [inline fussy.el:1087]: Maybe swap the or conditions to check if score is null first.

File: fussy.el [hunk 3]
```diff
@@ -1082,13 +1082,27 @@ highlighting."
...
+
+
 (defun fussy-fzf-native-score (str query &rest _args)
   "Score STR for QUERY using `fzf-native'."
   (require 'fzf-native)
   (when (fboundp 'fzf-native-score)
     (let ((str (funcall fussy-remove-bad-char-fn str))
           (query (fussy-encode-coding-string query)))
-      (fzf-native-score str query (fussy--fzf-native-slab)))))
+      (fussy--fzf-native-fix-score-indices
```
Comment (jojojames) [inline fussy.el:1103]: Wonder if it's possible for fzf-native to check this instead. @dangduc

Comment (dangduc) [reply fussy.el:1103]: I think an advice is better here. @buzztaiki I'd accept a PR to [fzf-native](https://github.com/dangduc/fzf-native).

```emacs-lisp
(defun fzf-native--fix-score-indices (fzf-native-score-fn &rest args)
  "Fix SCORE indices of `fzf-native-score' if STR is multibyte."
  (let ((res (apply fzf-native-score-fn args))
        (str (car args)))
    (if (or (not (multibyte-string-p str)) (null res))
        res
      ;; fzf-native makes score indices as byte position.
      ;; But we want it as character position.
      (let ((idx (cl-loop for c across str
                          for i = 0 then (1+ i)
                          vconcat (make-vector (string-bytes (char-to-string c)) i))))
        (cons (car res) (mapcar (lambda (x) (aref idx x)) (cdr res)))))))

(fzf-native-score "ãã±ã¢ã³.txt" "txt")
; => (80 25 26 27)

(advice-add 'fzf-native-score :around #'fzf-native--fix-score-indices)

(fzf-native-score "ãã±ã¢ã³.txt" "txt")
; => (80 13 14 15)
```

The PR fixes indices, but the actual scoring is also incorrect for multibyte. Due to character distance being a factor in scoring. And it looks like the distance calculation assumes chars are singlebyte. Meaning that multibyte candidates will score higher (or lower?) than they should in certain cases.

```emacs-lisp
(fzf-native-score "aaaatxt" "txt")
; => (56 4 5 6)

(fzf-native-score "asdftxt" "txt")
; => (56 4 5 6)

(fzf-native-score "txt" "txt")
; => (80 4 5 6)
```

Comment (buzztaiki) [reply fussy.el:1103]: > I think an advice is better here. @buzztaiki I'd accept a PR to [fzf-native](https://github.com/dangduc/fzf-native).

Ok, I'll send PR to fzf-native.

> The PR fixes indices, but the actual scoring is also incorrect for multibyte. Due to character distance being a factor in scoring. And it looks like the distance calculation assumes chars are singlebyte. Meaning that multibyte candidates will score higher (or lower?) than they should in certain cases.

I agree. I think we need to fix fzf.c to calculate the score correctly for multibyte characters.

Comment (buzztaiki) [reply fussy.el:1103]: @dangduc Which is better to use the advice or to call `fzf-native--fix-score-indices` in ` fzf_native_score`?

Comment (dangduc) [reply fussy.el:1103]: I would take a PR to fzf-native that adds the advice function `fzf-native--fix-score-indices`. Then all you'd need to do is to configure the advice. For example if using straight.el:
```diff
; package configuration file
(use-package fzf-native
  :ensure t
  :straight
  (fzf-native
   :repo "dangduc/fzf-native"
   :host github
   :files (:defaults "bin"))
  :config
+ (advice-add 'fzf-native-score :around #'fzf-native--fix-score-indices)
  (setq fussy-score-fn 'fussy-fzf-native-score)
 (fzf-native-load-dyn))
 ```

Alternatively the PR could just enable the advice by default in fzf-native.el package. So you wouldn't need to do `advice-add` in package configuration.

```diff
; fzf-native.el file
+(defun fzf-native--fix-score-indices (fzf-native-score-fn &rest args)
+  "Fix SCORE indices of `fzf-native-score' if STR is multibyte."
+  (let ((res (apply fzf-native-score-fn args))
+        (str (car args)))
+    (if (or (not (multibyte-string-p str)) (null res))
+        res
+      ;; fzf-native makes score indices as byte position.
+      ;; But we want it as character position.
+      (let ((idx (cl-loop for c across str
+                          for i = 0 then (1+ i)
+                          vconcat (make-vector (string-bytes (char-to-string c)) i))))
+        (cons (car res) (mapcar (lambda (x) (aref idx x)) (cdr res)))))))
+
+ (advice-add 'fzf-native-score :around #'fzf-native--fix-score-indices)
```

An alternative approach to advising is have the dynamic module code call `fzf-native--fix-score-indices`. But I think we want to limit calls to the emacs process from the dynamic module.

Comment (dangduc) [reply fussy.el:1103]: In increasing desirability, I'd want the fix to be:

1. Elisp code that fixes the indices, e.g. @buzztaiki's original PR or my suggestion to fzf-native.el.
2. C code that fixes the indices.
3. fzf.c fixes the indices.
4. fzf.c fixes the scoring + indices.

Comment (buzztaiki) [reply fussy.el:1103]: Thanks, I will enable the advice by default.

Comment (buzztaiki) [reply fussy.el:1103]: I've send PR in https://github.com/dangduc/fzf-native/pull/15

**Comment by jojojames** [comment]
This PR is not necessary anymore right? Closing, thanks for the report and the fix.

**Pull request closed without merging**